File: Either.cpp
#include <iostream>
using std::string;

template<class A, class B>class Either{
    bool    _isRight;
    A       left;
    B       right;

public:

    Either(A x) { left = x;  _isRight = false;}
    Either(B x) { right = x; _isRight = true;}

    void Left(A x) { left = x; _isRight = false;}
    void Right(B x) { right = x; _isRight = true;}

    bool isRight(){ return _isRight; } 
    bool isLeft(){ return !_isRight; } 

    A    getLeft() { return left;}
    B    getRight() { return right;}
};

int i(int n) {return n;}
int j(bool b) {return b ? 0:1;}

int m(Either<string, int> const & e) {

}


int main(int argc, char** argv) {

    auto eith = Either<string, int>("haha");

    if (argc >= 2) {
        eith.Left("haha");
    } else {
        eith.Right(42);
    }

    if (eith.isRight()) {
        std::cout << eith.getRight() << std::endl;
    } else {
        std::cout << eith.getLeft() << std::endl;
    }

}

File: Either.rs

#[derive(Debug)]
pub enum Either<T1, T2> {
    Left(T1),
    Right(T2),
}

impl<T1, T2> Either<T1, T2> {

    pub fn left(mut self, l: T1) -> Self {
        self = Either::Left(l);
        self
    }

    pub fn isLeft(&self) -> bool {
        match self {
            Either::Left(_) => true,
            _ => false,
        }
    }

    pub fn right(mut self, r: T2) -> Self {
        self = Either::Right(r);
        self
    }

    pub fn isRight(&self) -> bool {
        match self {
            Either::Right(_) => true,
            _ => false,
        }
    }
}

fn main () {
    let e: Either<i32, f32> = Either::Left(5);
    println!("{:?}", e.isRight());
    let e = e.right(5.0);
    println!("{:?}", e.isRight());
}
